### PR TITLE
Enhancing UX by introducing User Profile Modification Feature

### LINKED ISSUE #2261: User should be able to edit profile
**Is your feature request related to a problem? Please describe.**

Currently there is no functionality for user to edit its profile. 

There should be a proper functionality where user is able to change its profile picture, first name, last name and email address.  

**Describe the solution you'd like**
Proper Implementation of logic in edit_profile_view_model.dart file to update/edit user information.

**Additional context**
Screenshot_1703078288

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by Dante291**
**What kind of change does this PR introduce?**

Introducing user profile modification feature where user would be able to customize their profile.

**Issue Number:**

Fixes #2261 

**Video**

https://github.com/PalisadoesFoundation/talawa/assets/103507835/4427f69e-ab5e-4f79-8583-9e8302432cc1

**Did you add tests for your changes?**

Yes

**Does this PR introduce a breaking change?**

Yes

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by Dante291** [comment]
Made profile pic visible at other screens of the app too:

Screenshot_1703412506


Screenshot_1703412514

**Comment by Dante291** [comment]
@noman2002 @Cioppolo14 @Kevoniat Please assign reviewers for this, thanks.

File: lib/view_model/after_auth_view_models/add_post_view_models/add_post_view_model.dart [hunk 1]
```diff
@@ -75,9 +75,14 @@ class AddPostViewModel extends BaseModel {
   /// None
   /// returns:
   /// * `String`: The username of the currentUser
-  String get userName =>
+  String? get userName =>
```
Comment (literalEval) [inline lib/view_model/after_auth_view_models/add_post_view_models/add_post_view_model.dart:78]: Why mark this optional ?

File: lib/view_model/after_auth_view_models/add_post_view_models/add_post_view_model.dart [hunk 2]
```diff
@@ -75,9 +75,14 @@ class AddPostViewModel extends BaseModel {
...
       userConfig.currentUser.firstName! + userConfig.currentUser.lastName!;
 
+  /// User profile picture.
+  ///
+  /// more_info_if_required
```
Comment (literalEval) [inline lib/view_model/after_auth_view_models/add_post_view_models/add_post_view_model.dart:83]: Remove the 2nd and 3rd line. Please read Dart documentation guidelines once.

File: lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart [hunk 3]
```diff
@@ -61,6 +69,147 @@ class EditProfilePageViewModel extends BaseModel {
     }
   }
 
+  /// modal sheet to choose image.
+  ///
+  /// more_info_if_required
```
Comment (literalEval) [inline lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart:74]: Remove this line.

File: lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart [hunk 4]
```diff
@@ -61,6 +69,147 @@ class EditProfilePageViewModel extends BaseModel {
...
+      return base64Image!;
+    } catch (error) {
+      return '';
+    }
+  }
+
+  /// Method to update user profile.
+  ///
+  /// **params**:
+  /// * `firstName`: define_the_param
```
Comment (literalEval) [inline lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart:144]: Define what the parameter means. Don't leave it as is.

File: lib/views/after_auth_screens/add_post_page.dart [hunk 5]
```diff
@@ -81,8 +82,14 @@ class _AddPostState extends State<AddPost> {
...
               children: <Widget>[
                 ListTile(
-                  leading: const CircleAvatar(radius: 25),
-                  title: Text(model.userName),
+                  leading: CustomAvatar(
+                    isImageNull: model.userPic == null,
+                    firstAlphabet:
+                        model.userName!.substring(0, 1).toUpperCase(),
+                    imageUrl: model.userPic,
+                    fontSize: 20,
```
Comment (literalEval) [inline lib/views/after_auth_screens/add_post_page.dart:90]: Don't use constant sizing. Use SizeConfig instead.

File: lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart [hunk 6]
```diff
@@ -61,6 +69,147 @@ class EditProfilePageViewModel extends BaseModel {
...
+  ///
+  /// **params**:
+  /// * `context`: define_the_param
+  ///
+  /// **returns**:
+  ///   None
+  void showImagePickerIcons(BuildContext context) {
+    showModalBottomSheet(
+      context: context,
+      builder: (BuildContext bc) {
```
Comment (literalEval) [inline lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart:84]: Rename this to `context`.

File: lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart [hunk 7]
```diff
@@ -61,6 +69,147 @@ class EditProfilePageViewModel extends BaseModel {
...
+  /// modal sheet to choose image.
+  ///
+  /// more_info_if_required
+  ///
+  /// **params**:
+  /// * `context`: define_the_param
+  ///
+  /// **returns**:
+  ///   None
+  void showImagePickerIcons(BuildContext context) {
```
Comment (literalEval) [inline lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart:81]: If it is an ImagePicker Dialog, then why is it named showImagePickerIcons ?
Moreover, passing around contexts in functions is not the standard way to go about this. Please have a look at MVVM architecture once.

File: lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart [hunk 8]
```diff
@@ -61,6 +69,147 @@ class EditProfilePageViewModel extends BaseModel {
...
+  }
+
+  /// This function is used to convert the image into Base64 format.
+  ///
+  /// **params**:
+  /// * `file`:  Takes the image in format of file.
+  ///
+  /// **returns**:
+  /// * `Future<String>`: image in string format
+  Future<String> convertToBase64(File file) async {
```
Comment (literalEval) [inline lib/view_model/after_auth_view_models/profile_view_models/edit_profile_view_model.dart:131]: You'll need to remove this in another PR once the ImageService one gets merged.

**Review by literalEval — CHANGES_REQUESTED** [review]
1. Remove print statements.
2. Write good documentation. Don't leave the template as is.
3. Add empty lines to make the code more readable.

**Comment by Dante291** [comment]
@literalEval Sir I have resolved all the requested changes except the constant sizing one as constant sizing is being used for fontsize and icon size throughout the codebase, constant sizing makes more sense here as sizeConfig utilizes a dynamic approach by comparing with the entire screen resolution, so user had to try number of times in this case, do correct me if I'm wrong here.

other than that, I have used sizeConfig for sizing for all widgets and all.

**Comment by literalEval** [comment]
@Dante291 fix the failing checks. Refactor the code wherever necessary.

**Comment by Dante291** [comment]
> @Dante291 fix the failing checks. Refactor the code wherever necessary.

Sir as suggested by you to follow MVVM architecture, I moved the code for modal sheet in the view file but due to this limit of 300 lines of code is exceeded, as view files contain whole UI logic of the screen there shouldnt be limit for view files, profile_page.dart file also has like 500+ lines so this check could be ignored here too.

**Comment by Dante291** [comment]
@literalEval please review sir.

File: lib/views/after_auth_screens/profile/edit_profile_page.dart [hunk 9]
```diff
@@ -317,4 +282,29 @@ class _EditProfilePageState extends State<EditProfilePage> {
...
+  /// **params**:
+  /// * `context`:context for the sheet
+  /// * `icon`: icon for the method
+  /// * `label`: label for the method
+  /// * `onTap`: onTap funtionality for the method
+  ///
+  /// **returns**:
+  /// * `Widget`: Icon Button for selecting different image selection method.
+  Widget _createModalSheetButton(
+      BuildContext context, IconData icon, String label, VoidCallback onTap) {
```
Comment (literalEval) [inline lib/views/after_auth_screens/profile/edit_profile_page.dart]: Add a trailing comma and reformat.

File: lib/views/after_auth_screens/profile/edit_profile_page.dart [hunk 10]
```diff
@@ -317,4 +282,29 @@ class _EditProfilePageState extends State<EditProfilePage> {
...
+    return GestureDetector(
+      key: Key('select$label'),
+      onTap: onTap,
+      child: Column(
+        mainAxisSize: MainAxisSize.min,
+        children: <Widget>[
+          Icon(icon, size: SizeConfig.screenHeight! * 0.05),
```
Comment (literalEval) [inline lib/views/after_auth_screens/profile/edit_profile_page.dart:295]: Add trailing comma and reformat lines 304 and 305.

**Review by literalEval — CHANGES_REQUESTED** [review]

File: lib/views/after_auth_screens/profile/profile_page.dart [hunk 11]
```diff
@@ -63,9 +63,8 @@ class ProfilePage extends StatelessWidget {
                     context: context,
                     builder: (BuildContext context) {
                       return Container(
-                        height: 200,
+                        height: 150,
```
Comment (literalEval) [inline lib/views/after_auth_screens/profile/profile_page.dart:66]: Use SizeConfig.

File: lib/views/after_auth_screens/profile/profile_page.dart [hunk 12]
```diff
@@ -118,191 +116,198 @@ class ProfilePage extends StatelessWidget {
...
-                          children: [
-                            RaisedRoundedButton(
-                              key: homeModel!.keySPDonateUs,
-                              buttonLabel:
-                                  AppLocalizations.of(context)!.strictTranslate(
-                                'Donate to the Community',
+                            Expanded(
+                              flex: 1,
+                              child: IconButton(
+                                key: const Key('inv1'),
```
Comment (literalEval) [inline lib/views/after_auth_screens/profile/profile_page.dart:164]: Use a better key name.

**Review by literalEval — CHANGES_REQUESTED** [review]

**Comment by literalEval** [comment]
@Dante291 use SizeConfig everywhere. Similarly, add trailing commas wherever needed and refactor.

**Comment by Dante291** [comment]
@literalEval properly formatted and used dynamic sizing wherever possible, please review sir.

**Review by literalEval — CHANGES_REQUESTED** [review]
Cover the uncovered lines in tests. Use codecov to find out the missing lines.

**Comment by Dante291** [comment]
> ## [Codecov](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2268?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation) Report
> All modified and coverable lines are covered by tests ✅
> 
> > Comparison is base [(`c902baa`)](https://app.codecov.io/gh/PalisadoesFoundation/talawa/commit/c902baaf1467b985b5301a6b30a06d92516e4956?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation) 88.27% compared to head [(`a459cb5`)](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2268?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation) 90.08%.
> > Report is 4 commits behind head on develop.
> 
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2268?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation).

@literalEval All lines covered sir

**Review by literalEval — APPROVED** [review]
LGTM

**Review by palisadoes — APPROVED** [review]

**Pull request merged**